File: neetcode-150/Arrays_and_hashing/longest_consecutive_sequence/solution.js
let nums = [100, 4, 200, 1, 3, 2, 5, 6, 7, 8];

function longestConsecutive(nums) {
    let sol = new Solution();
    return sol.longestConsecutive(nums);
}

class Solution {
    longestConsecutive(nums) {
        // Convert to Set for O(1) lookup
        const num_set = new Set(nums);

        let longest = 0;

        // Check only if it's the start of a sequence
        for (let num of num_set) {
            if (!num_set.has(num - 1)) {
                let current_streak = 1;

                // Count consecutive numbers
                while (num_set.has(num + 1)) {
                    num++;
                    current_streak++;
                }

                // Update max streak
                longest = Math.max(longest, current_streak);
            }
        }

        return longest;
    }
}

console.log(longestConsecutive(nums));

File: neetcode-150/Arrays_and_hashing/longest_consecutive_sequence/solution.py
from typing import List

nums = [100, 4, 200, 1, 3, 2, 5, 6, 7, 8]

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # Convert to set for O(1) lookups
        num_set = set(nums)

        longest = 0

        # Check if num starts a sequence
        for n in num_set:
            if n - 1 not in num_set:
                current_streak = 1
                while n + 1 in num_set:
                    n += 1
                    current_streak += 1
                longest = max(longest, current_streak)

        return longest

sol = Solution()
print(sol.longestConsecutive(nums))
